pycqed/measurement/waveform_control/circuit_builder.py

```diff
@@ -155,39 +157,117 @@ def get_cz_operation_name(self, qb1=None, qb2=None, op_code=None, **kw):
+        it has to be in the order 'sm'.
+
         Args:
-            op: operation
-            parse_z_gate: whether or not to look for Zgates with arbitrary
-            angles.
+            op: operation (str in the above format, or iterable
+            corresponding to the splitted string)
+            parse_rotation_gates: whether or not to look for gates with
+            arbitrary angles.
 
         Returns: deepcopy of the pulse dictionary
 
         """
-        op_info = op.split(" ")
+        op_info = op.split(" ") if isinstance(op, str) else op
         # the call to get_qubits resolves qubits indices if needed
         _, op_info[1:] = self.get_qubits(op_info[1:])
-        op = op_info[0] + ' ' + ' '.join(op_info[1:])
-        if parse_z_gate and op.startswith("Z"):
-            # assumes operation format of f"Z{angle} qbname"
-            # FIXME: This parsing is format dependent and is far from ideal but
-            #  until we can get parametrized pulses it is helpful to be able to
-            #  parse Z gates
-            angle, qbn = op_info[0][1:], op_info[1]
-            p = self.get_pulse(f"Z180 {qbn}", parse_z_gate=False)
-            p['basis_rotation'] = {qbn: float(angle)}
-        elif op_info[0].startswith('CZ'):
+        op_name = op_info[0][1:] if op_info[0][0] == 's' else op_info[0]
+        op = op_name + ' ' + ' '.join(op_info[1:])
+
+        if op_name.startswith('CZ'):
             operation = self.get_cz_operation_name(op_info[1], op_info[2])
             p = deepcopy(self.operation_dict[operation])
+        elif parse_rotation_gates and op not in self.operation_dict:
+            # assumes operation format of, e.g., f" Z{angle} qbname"
+            # FIXME: This parsing is format dependent and is far from ideal but
+            #  to generate parametrized pulses it is helpful to be able to
+            #  parse Z gates etc.
+            factor = -1 if op_name[0] == 'm' else 1
+            if factor == -1:
+                op_name = op_name[1:]
+            if op_name[0] not in ['X', 'Y', 'Z']:
+                raise KeyError(f'Gate "{op}" not found.')
+            angle, qbn = op_name[1:], op_info[1]
+            param = None
+            if angle[0] == ':':  # angle depends on a parameter
+                angle = angle[1:]
+                param_start = angle.find('[') + 1
+                # If '[' is contained, this indicates that the parameter
+                # is part of a mathematical expression. Otherwise, the angle
+                # is equal to the parameter.
+                if param_start > 0:
+                    param_end = angle.find(']', param_start)
+                    param = angle[param_start:param_end]
+                    angle = angle.replace('[' + param + ']', 'x')
+                    f = eval('lambda x : ' + angle)
+                else:
+                    param = angle
+
+            if not self.decompose_rotation_gates.get(op_name[0], False):
```

Comment:
I always find it more readable to avoid double negative of the default behavior, i.e. : 
``` 
if self.decompose_rotation_gates.get(op_name[0], False):
    raise NotImplementedError()
else: 
    ...
```